include/deal.II/lac/constraint_matrix.templates.h

```diff
@@ -2319,10 +2330,42 @@ ConstraintMatrix::distribute_local_to_global (
                                                    local_matrix);
 
           if (val != number ())
-            global_vector(row) += static_cast<typename VectorType::value_type>(val);
+            {
+              rows[local_row_n] = row;
+              vector_values[local_row_n] = val;
+              ++local_row_n;
+            }
         }
     }
 
+  const size_type n_local_rows = local_row_n;
+  // Attempt to use the faster code path for PETSc if we are dealing with
+  // PetscScalars.
+#ifdef DEAL_II_WITH_PETSC
+  PETScWrappers::VectorBase *petsc_global_vector
+    = dynamic_cast<PETScWrappers::VectorBase *>(&global_vector);
+  if (petsc_global_vector != 0 && types_are_equal<PetscScalar, number>::value)
+    {
+      rows.resize(n_local_rows);
+      vector_values.resize(n_local_rows);
```

Comment:
Can you add a comment before these lines of the kind "Drop elements of the 'rows' and 'vector_values' objects that were not used because we don't write zero elements" to make it clear that you're resizing to a smaller size, dropping unused elements.